modules/core/src/main/java/org/apache/ignite/internal/processors/cache/QueryCursorImpl.java

```diff
@@ -142,14 +152,37 @@ public QueryCursorImpl(Iterable<T> iterExec, GridQueryCancel cancel, boolean isQ
                 clo.consume(iter.next());
         }
         finally {
+            // Update state if the results is read to end
+            STATE_UPDATER.compareAndSet(this, RESULT_READY, NO_DATA);
+
             close();
         }
     }
 
     /** {@inheritDoc} */
     @Override public void close() {
         while (state != CLOSED) {
+            // Check that iterator has no data: in this case cancel.cancel() shouldn't be called.
+            try {
+                if (iter != null && !iter.hasNext())
+                    STATE_UPDATER.set(this, NO_DATA);
+            }
+            catch (Exception e) {
+                // Ignore exception on check iterator
+                // because Iterator.hasNext() may throw error on invalid / error query.
+                STATE_UPDATER.set(this, NO_DATA);
+            }
+
+            if (STATE_UPDATER.compareAndSet(this, NO_DATA, CLOSED)) {
+                closeIter();
```

Comment:
potentially Iterator can be closed twice in case we have two concurrent ``close`` invocations here, because earlier you  set ``NO_DATA`` without any doubts